File: go/31_graph/graph_search.go
package graph

import (
	"container/list"
	"fmt"
)

//adjacency table, 无向图
type Graph struct {
	adj []*list.List
	v   int
}

//init graphh according to capacity
func newGraph(v int) *Graph {
	graphh := &Graph{}
	graphh.v = v
	graphh.adj = make([]*list.List, v)
	for i := range graphh.adj {
		graphh.adj[i] = list.New()
	}
	return graphh
}

//insert as add edge，一条边存2次
func (self *Graph) addEdge(s int, t int) {
	self.adj[s].PushBack(t)
	self.adj[t].PushBack(s)
}

//search path by BFS
func (self *Graph) BFS(s int, t int) {

	//todo
	if s == t {
		return
	}

	//init prev
	prev := make([]int, self.v)
	for index := range prev {
		prev[index] = -1
	}

	//search by queue
	var queue []int
	visited := make([]bool, self.v)
	queue = append(queue, s)
	visited[s] = true
	isFound := false
	for len(queue) > 0 && !isFound {
		top := queue[0]
		queue = queue[1:]
		linkedlist := self.adj[top]
		for e := linkedlist.Front(); e != nil; e = e.Next() {
			k := e.Value.(int)
			if !visited[k] {
				prev[k] = top
				if k == t {
					isFound = true
					break
				}
				queue = append(queue, k)
				visited[k] = true
			}
		}
	}

	if isFound {
		printPrev(prev, s, t)
	} else {
		fmt.Printf("no path found from %d to %d\n", s, t)
	}

}

//search by DFS
func (self *Graph) DFS(s int, t int) {

	prev := make([]int, self.v)
	for i := range prev {
		prev[i] = -1
	}

	visited := make([]bool, self.v)
	visited[s] = true

	isFound := false
	self.recurse(s, t, prev, visited, isFound)

	printPrev(prev, s, t)
}

//recursivly find path
func (self *Graph) recurse(s int, t int, prev []int, visited []bool, isFound bool) {

	if isFound {
		return
	}

	visited[s] = true

	if s == t {
		isFound = true
		return
	}

	linkedlist := self.adj[s]
	for e := linkedlist.Front(); e != nil; e = e.Next() {
		k := e.Value.(int)
		if !visited[k] {
			prev[k] = s
			self.recurse(k, t, prev, visited, false)
		}
	}

}

//print path recursively
func printPrev(prev []int, s int, t int) {

	if t == s || prev[t] == -1 {
		fmt.Printf("%d ", t)
	} else {
		printPrev(prev, s, prev[t])
		fmt.Printf("%d ", t)
	}

}

//func main() {
//	graph := newGraph(8)
//	graph.addEdge(0, 1)
//	graph.addEdge(0, 3)
//	graph.addEdge(1, 2)
//	graph.addEdge(1, 4)
//	graph.addEdge(2, 5)
//	graph.addEdge(3, 4)
//	graph.addEdge(4, 5)
//	graph.addEdge(4, 6)
//	graph.addEdge(5, 7)
//	graph.addEdge(6, 7)
//
//	graph.BFS(0, 7)
//	fmt.Println()
//	graph.BFS(1, 3)
//	fmt.Println()
//	graph.DFS(0, 7)
//	fmt.Println()
//	graph.DFS(1, 3)
//	fmt.Println()
//}

File: rust/31_graph/graph_search.rs
use std::collections::LinkedList;
use std::collections::VecDeque;

// 无向图
#[derive(Debug)]
struct Graph {
    v:   i32,
    linked_vec: Vec<LinkedList<i32>>, // 邻接表
}

impl Graph {
    fn new(v: i32) -> Self {
        Graph {
            v: v,
            linked_vec: vec![LinkedList::new(); v as usize],
        }
    }

    // 无向图的每条边存两次
    fn add_edge(&mut self, s: i32, t: i32) {
        self.linked_vec[s as usize].push_back(t);
        self.linked_vec[t as usize].push_back(s);
    }

    fn bfs(&self, s: i32, t: i32) {
        if s == t { return; }

        let mut prev    = vec![-1; self.v as usize];
        let mut visited = vec![false; self.v as usize];
        let mut queue   = VecDeque::new();

        visited[s as usize] = true;
        queue.push_back(s);

        while !queue.is_empty() {
            let w = queue.pop_front().unwrap();
            for item in self.linked_vec[w as usize].iter() {
                if visited[*item as usize] { continue; }
                prev[*item as usize] = w;
                if *item == t {
                    self.draw(&prev, s, t);
                    return;
                }
                visited[*item as usize] = true;
                queue.push_back(*item);
            }
        }
    }

    fn dfs(&self, s: i32, t: i32) {
        let mut found   = false;
        let mut prev    = vec![-1; self.v as usize];
        let mut visited = vec![false; self.v as usize];

        self.recur_dfs(s, t, &mut visited, &mut prev, &mut found);
        self.draw(&prev, s, t);
    }

    fn recur_dfs(&self,
                 s:       i32,
                 t:       i32,
                 visited: &mut Vec<bool>,
                 prev:    &mut Vec<i32>,
                 found:   &mut bool) {
        if *found == true { return; }
        visited[s as usize] = true;
        if s == t {
            *found = true;
            return;
        }
        for item in self.linked_vec[s as usize].iter() {
            if visited[*item as usize] { continue; }
            prev[*item as usize] = s;
            self.recur_dfs(*item, t, visited, prev, found);
        }
    }

    // 递归打印路径
    fn draw(&self, prev: &Vec<i32>, s: i32, t: i32) {
        if prev[t as usize] != -1 && s != t {
            self.draw(prev, s, prev[t as usize]);
        }

        println!("{} ->", t);
    }
}

fn main() {
    let mut graph = Graph::new(8);
    graph.add_edge(0, 1);
    graph.add_edge(0, 3);
    graph.add_edge(1, 2);
    graph.add_edge(1, 4);
    graph.add_edge(2, 5);
    graph.add_edge(3, 4);
    graph.add_edge(4, 5);
    graph.add_edge(4, 6);
    graph.add_edge(5, 7);
    graph.add_edge(6, 7);

    // Graph { v: 8, linked_vec: [[1, 3], [0, 2, 4], [1, 5], [0, 4], [1, 3, 5, 6], [2, 4, 7], [4, 7], [5, 6]] }
    println!("{:?}", graph);
    graph.bfs(0, 7);
    println!("bfs=============");
    graph.bfs(1, 3);
    println!("bfs=============");

    graph.dfs(0, 7);
    println!("dfs=============");
    graph.dfs(1, 3);
    println!("dfs=============");
}
